Modules/HTTP/Sources/HTTP/Parser/MessageParser.swift

```diff
@@ -0,0 +1,316 @@
+                    return String(cString: ptr.baseAddress!)
+                }
+                context.currentHeaderField = CaseInsensitiveString(string)
+                
+                
+            case .headerValue:
+                buffer.append(0)
+                let string = buffer.withUnsafeBufferPointer { (ptr: UnsafeBufferPointer<UInt8>) -> String in
+                    return String(cString: ptr.baseAddress!)
+                }
+                context.addValueForCurrentHeaderField(string)
+                
+            case .headersComplete:
+                context.currentHeaderField = nil
+                context.method = Request.Method(code: http_method(rawValue: parser.method))
+                context.version = Version(major: Int(parser.http_major), minor: Int(parser.http_minor))
+                
+            case .body:
+                context.body = Buffer(buffer)
+                
+            }
+            
+            buffer = []
+            state = newState
+            
+            if state == .messageComplete {
+                let message: Message
+                switch mode {
+                case .request:
+                    var request = Request(method: context.method!,
+                                          url: context.url!,
+                                          headers: Headers(),
+                                          body: .buffer(context.body))
+                    request.headers = Headers(context.headers)
+                    
+                    message = request
+                case .response:
+                    let cookieHeaders =
+                        self.context.headers
+                            .filter { $0.key == "Set-Cookie" }
+                            .map { $0.value }
+                            .reduce(Set<String>()) { initial, value in
+                                return initial.union(Set(value.components(separatedBy: ", ")))
+                            }
+                    
+                    let response = Response(version: context.version!,
+                                            status: context.status!,
+                                            headers: Headers(context.headers),
+                                            cookieHeaders: cookieHeaders,
+                                            body: .buffer(context.body))
+                    
+                    message = response
+                }
+                
+                messages.append(message)
+                context = Context()
+            }
+        }
+        
+        
+        
+        guard let data = data, data.count > 0 else {
+            return
+        }
+        
+        data.baseAddress!.withMemoryRebound(to: UInt8.self, capacity: data.count) { ptr in
+            for i in 0..<data.count {
+                self.buffer.append(ptr[i])
+            }
+        }
+    }
+    
+}
+
+extension MessageParser.Error: Error, CustomStringConvertible {
```

Comment:
I think I prefer `MessageParserError` instead of `MessageParser.Error` to keep consistency with everything else. It would be just changing from:

```swift
class MessageParser {
    typealias Error = http_errno
}
```

to:

```swift
typealias MessageParserError = http_errno
```

To clarify, the reason we prefer to put the errors outside of the type is because generic types don't allow nested types. So for consistency we always declare the error outside the type.